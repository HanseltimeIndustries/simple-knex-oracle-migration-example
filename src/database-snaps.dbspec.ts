import Knex, { Knex as KnexType} from 'knex'
import { join } from 'path'

let client: KnexType

beforeAll(async () => {
    const config = await import(join(__dirname, '..', 'knexfile.js'))

    const envConfig = config['dbTest']
    client = Knex(envConfig)

    const migrationTable = envConfig.tableName ?? 'knex_migrations'
    // check if there is a migrations table
    const table = await client.raw(`
        SELECT table_name FROM user_tables
        WHERE table_name = '${migrationTable}'
    `)

    if (table.length > 0) {
        // Ensure that this database is not already half-migrated
        const existingMigrations = await client.raw(`
            SELECT COUNT(*) as cnt FROM "${migrationTable}"
            `)
        const count = existingMigrations[0].CNT
        if (count > 0) {
            throw new Error(`Attempting to perform database migration snapshots on a non-clean database.  Please start with a fresh DB.`)
        }
    }


    // Run all migrations
    await client.migrate.latest()
})

afterAll(async () => {
    await client?.destroy()
})

it('matches snapshots for the db stored_procedures', async () => {
    const sources = await client<any>('ALL_SOURCE')
        .select('TEXT')
        .where({
            TYPE: 'PROCEDURE',
            OWNER: 'SYSTEM'
        })

    const procedures = parseTextFiles(sources)
    
    procedures.forEach((procedure) => {
        expect(procedure.source).toMatchSnapshot(procedure.name)
    })
})

function parseTextFiles(sourceLines: {
    TEXT: string
}[]) {
    let idx = -1
    return sourceLines.reduce((sources, sourceLine) => {
        const textPiece = sourceLine.TEXT
        if (textPiece.startsWith('PROCEDURE ')) {
            idx++
            const match = /PROCEDURE\s+[^\s]+/.exec(textPiece)
            sources.push({
                name: match![0],
                source: textPiece
            })
        } else {
            sources[idx].source += textPiece
        }
        return sources
    }, [] as {
        name: string
        source: string
    }[])
}