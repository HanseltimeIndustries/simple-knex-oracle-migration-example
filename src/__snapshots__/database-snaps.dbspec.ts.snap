// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshots for the db stored_procedures: PROCEDURE "SelectAllDiscisples" 1`] = `
"PROCEDURE "SelectAllDiscisples"
    AS
    BEGIN
    SELECT * FROM kungfu_disciple WHERE id = 77;
    END;
"
`;

exports[`matches snapshots for the db stored_procedures: PROCEDURE SelectAllDiscisples 1`] = `
"PROCEDURE SelectAllDiscisples
    AS
    BEGIN
    SELECT * FROM kungfu_disciple;
    END
"
`;
